### PR TITLE
Fix late static binding calls

### PR DESCRIPTION
**Pull request by mvorisek**
fix https://github.com/phpstan/phpstan/issues/10469

based on https://github.com/phpstan/phpstan/issues/10469#issuecomment-1906065516 analysis

### CODE REVIEW

**Comment by mvorisek** [comment]
@ondrejmirtes might I ask you to review this fix

File: src/Analyser/MutatingScope.php [hunk 1]
```diff
@@ -5637,6 +5637,21 @@ private function methodCallReturnType(Type $typeWithMethod, string $methodName,
 		}
 
 		$methodReflection = $typeWithMethod->getMethod($methodName, $this);
+
+		// https://www.php.net/manual/en/language.oop5.late-static-bindings.php
+		if (
+			$methodCall instanceof Expr\StaticCall
+			&& $typeWithMethod instanceof StaticType
```
Comment (ondrejmirtes) [inline src/Analyser/MutatingScope.php]: This isn't a proper fix. The place where the StaticType transformation gets decided is in `ObjectType::getUnresolvedMethodPrototype()` and `StaticType::getUnresolvedMethodPrototype()`. One of them transforms `static` to the class it's called on, and the other one preserves `static`. Some piece of logic should be written there.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Analyser/MutatingScope.php (hunk 1 above)
Comment (mvorisek) [reply src/Analyser/MutatingScope.php]: But how - in `ObjectType::getUnresolvedMethodPrototype()` there is no info about how the call was made - ie. `$methodCall` and `$typeWithMethod` is not passed. Maybe you can look into the fix with the tests below?

Comment (ondrejmirtes) [reply src/Analyser/MutatingScope.php]: What kind of call do you want to transform and what kind of call do you not want to transform?

Comment (mvorisek) [reply src/Analyser/MutatingScope.php]: See condition a few lines below - `['self', 'static', 'parent']` - even if `self` resolves to the same class where declared, it does not behave the same due php "late static binding".

Comment (ondrejmirtes) [reply src/Analyser/MutatingScope.php]: I've asked for examples of method calls.

Comment (mvorisek) [reply src/Analyser/MutatingScope.php]: see the tests and/or https://github.com/phpstan/phpstan/issues/10469

Comment (ondrejmirtes) [reply src/Analyser/MutatingScope.php]: Yes, this comment explains it as I need it: https://github.com/phpstan/phpstan/issues/10469#issuecomment-1906046819

Comment (mvorisek) [reply src/Analyser/MutatingScope.php:5644]: Thanks - now the question is how to fix it.

**Comment by ondrejmirtes** [comment]
This works for me. Thank you.

**Pull request merged**